File: OOP/lab7.py
# Turchanov Denis 4PM
"""Задание паттерн Repository
1. Создать шаблонный интерфейс IRepository<T>:
void Add(T* item)
• void Update(T" item)
void Delete(T* item)
std::vector<T> Get(const std::string& where, const std::string& orderBy)
2. Создать класс User.
• int mld;
⚫ std::string mName
• std::string mEmail
• std::string mPhone
• std::string mCity
• Gender (Male, Female) mGender Добавить геттеры и сеттеры
3. Создать интерфейс UserRepository: public IRepository<User>:
•T getById(int id);
• T getByname(const std::string& name)
T getByEmail(const std::string& email)
• std::vector<T> getByCity (const std::string& city)
• std::vector<T> getByGender(Gender gender)
4. Создать класс Memory Repository<T>: public IRepository<Т», который реализует работу с объектами в оперативной памяти
5. Создать класс Memory UserRepository: public MemoryRepository<T>, public IUserRepository который реализует работу с объектами User в оперативной памяти
6. Продемонстрировать работу приложения при помощи добавления. изменения, удаления и получения объектов User из репозитория Memory UserRepository"""

from enum import Enum

class Gender(Enum):
    Male = "Male"
    Female = "Female"

class User:
    def __init__(self, user_id, name, email, phone, city, gender):
        self.user_id = user_id
        self.name = name
        self.email = email
        self.phone = phone
        self.city = city
        self.gender = gender

class IRepository:
    def add(self, item):
        pass

    def update(self, item):
        pass

    def remove(self, item):
        pass

    def get(self, where, order_by):
        pass

class IUserRepository(IRepository):
    def get_by_id(self, user_id):
        pass

    def get_by_name(self, name):
        pass

    def get_by_email(self, email):
        pass

    def get_by_city(self, city):
        pass

    def get_by_gender(self, gender):
        pass

class MemoryRepository(IRepository):
    def __init__(self):
        self.items = []

    def add(self, item):
        self.items.append(item)
        print(f"Added: {item.__dict__}")

    def update(self, item):
        for i, existing_item in enumerate(self.items):
            if existing_item.user_id == item.user_id:
                self.items[i] = item
                print(f"Updated: {item.__dict__}")
                break

    def remove(self, item):
        self.items.remove(item)
        print(f"Removed: {item.__dict__}")

    def get(self, where, order_by):
        return self.items

class MemoryUserRepository(MemoryRepository, IUserRepository):
    def get(self):
        users = []
        for user in self.items:
            users.append(user)
        return users

    def get_by_id(self, user_id):
        for user in self.items:
            if user.user_id == user_id:
                return user
        return None

    def get_by_name(self, name):
        for user in self.items:
            if user.name == name:
                return user
        return None

    def get_by_email(self, email):
        for user in self.items:
            if user.email == email:
                return user
        return None

    def get_by_city(self, city):
        return [user for user in self.items if user.city == city]

    def get_by_gender(self, gender):
        return [user for user in self.items if user.gender == gender]

# Пример использования
user_repository = MemoryUserRepository()

# Добавление пользователя
new_user = User(1, "John Doe", "john@example.com", "123-456-7890", "City", Gender.Male)
user_repository.add(new_user)

new_user1 = User(2, "John Doe1", "john@example.com1", "123-456-7890", "City", Gender.Male)
user_repository.add(new_user1)
# Изменение пользователя
updated_user = User(1, "John Doe Updated", "john@example.com", "123-456-7890", "City", Gender.Male)
user_repository.update(updated_user)

# Удаление пользователя
user_repository.remove(updated_user)

# Получение пользователей по различным условиям
users_in_city = user_repository.get_by_city("City")
print(f"Users in City: {users_in_city}")

female_users = user_repository.get_by_gender(Gender.Female)
print(f"Female Users: {female_users}")

File: OOP/lab7.cpp
// Turchanov Denis 4PM
// или на плюсах
#include <iostream>
#include <vector>
#include <algorithm>

enum class Gender { Male, Female };

class User {
public:
    int mId;
    std::string mName;
    std::string mEmail;
    std::string mPhone;
    std::string mCity;
    Gender mGender;

    // Геттеры и сеттеры

    int getId() const {
        return mId;
    }

    std::string getName() const {
        return mName;
    }

    std::string getEmail() const {
        return mEmail;
    }

    std::string getPhone() const {
        return mPhone;
    }

    std::string getCity() const {
        return mCity;
    }

    Gender getGender() const {
        return mGender;
    }

    void setId(int id) {
        mId = id;
    }

    void setName(const std::string& name) {
        mName = name;
    }

    void setEmail(const std::string& email) {
        mEmail = email;
    }

    void setPhone(const std::string& phone) {
        mPhone = phone;
    }

    void setCity(const std::string& city) {
        mCity = city;
    }

    void setGender(Gender gender) {
        mGender = gender;
    }
};

class IRepository {
public:
    virtual void add(User* item) = 0;
    virtual void update(User* item) = 0;
    virtual void remove(User* item) = 0;
    virtual std::vector<User> get(const std::string& where, const std::string& orderBy) = 0;
};

class IUserRepository : public IRepository {
public:
    virtual User getById(int id) = 0;
    virtual User getByName(const std::string& name) = 0;
    virtual User getByEmail(const std::string& email) = 0;
    virtual std::vector<User> getByCity(const std::string& city) = 0;
    virtual std::vector<User> getByGender(Gender gender) = 0;
};

class MemoryRepository : public IRepository {
protected:
    std::vector<User*> items;

public:
    virtual void add(User* item) override {
        items.push_back(item);
    }

    virtual void update(User* item) override {
        // Реализация обновления объекта
        // Найдем пользователя по ID и обновим его данные
        auto it = std::find_if(items.begin(), items.end(), [item](const User* u) {
            return u->getId() == item->getId();
        });

        if (it != items.end()) {
            // Обновим данные
            (*it)->setName(item->getName());
            (*it)->setEmail(item->getEmail());
            (*it)->setPhone(item->getPhone());
            (*it)->setCity(item->getCity());
            (*it)->setGender(item->getGender());
        }
    }

    virtual void remove(User* item) override {
        items.erase(std::remove(items.begin(), items.end(), item), items.end());
        delete item;
    }

    virtual std::vector<User> get(const std::string& where, const std::string& orderBy) override {
        // Реализация получения объектов с учетом условия и сортировки
        std::vector<User> result;
        for (auto item : items) {
            result.push_back(*item);
        }
        return result;
    }
};

class MemoryUserRepository : public MemoryRepository, public IUserRepository {
public:
    virtual User getById(int id) override {
        for (auto user : items) {
            if (user->getId() == id) {
                return *user;
            }
        }
        return User{};
    }

    virtual User getByName(const std::string& name) override {
        for (auto user : items) {
            if (user->getName() == name) {
                return *user;
            }
        }
        return User{};
    }

    virtual User getByEmail(const std::string& email) override {
        for (auto user : items) {
            if (user->getEmail() == email) {
                return *user;
            }
        }
        return User{};
    }

    virtual std::vector<User> getByCity(const std::string& city) override {
        std::vector<User> result;
        for (auto user : items) {
            if (user->getCity() == city) {
                result.push_back(*user);
            }
        }
        return result;
    }

    virtual std::vector<User> getByGender(Gender gender) override {
        std::vector<User> result;
        for (auto user : items) {
            if (user->getGender() == gender) {
                result.push_back(*user);
            }
        }
        return result;
    }
};

int main() {
    MemoryUserRepository userRepository;

    // Добавление пользователя
    User* newUser = new User{1, "John Doe", "john@example.com", "123-456-7890", "City", Gender::Male};
    userRepository.add(newUser);

    // Изменение пользователя
    User updatedUser{1, "John Doe Updated", "john@example.com", "123-456-7890", "City", Gender::Male};
    userRepository.update(&updatedUser);

    // Удаление пользователя
    userRepository.remove(&updatedUser);

    // Получение пользователей по различным условиям
    std::vector<User> usersInCity = userRepository.getByCity("City");
    std::vector<User> femaleUsers = userRepository.getByGender(Gender::Female);

    return 0;
}
